lib/multipipe/multi_pipe.go

```diff
@@ -0,0 +1,120 @@
+package multipipe
+
+import (
+	"io"
+	"sync"
+)
+
+// TODO: reduce lock contention if it becomes a bottleneck
+
+// MultiPipe is an io.Writer that can create multiple readers from its contents
+type MultiPipe struct {
+	buf    []byte
+	rdErr  error
+	closed bool
+	cond   *sync.Cond
+}
+
+// Reader is an io.Reader that reads from the beginning of its parent MultiPipe
+// without affecting the other readers
+type Reader struct {
+	parent *MultiPipe
+	offset int
+}
+
+// CopyPipe copies all the bytes from r to w without blocking, returns a channel
+// that will be closed on completion
+func CopyPipe(r io.Reader, w *MultiPipe) chan struct{} {
+	ch := make(chan struct{})
+
+	go func() {
+		_, err := io.Copy(w, r)
+		w.CloseWithError(err)
+		close(ch)
+	}()
+
+	return ch
+}
+
+// Read reads all the available contents from the MultiPipe parent, then if
+// there is a read error or the stream is closed, an error will be returned
+func (m *Reader) Read(p []byte) (n int, err error) {
+	if len(p) == 0 {
+		return 0, nil
+	}
+
+	m.parent.cond.L.Lock()
+	defer m.parent.cond.L.Unlock()
+
+	if !m.parent.closed || m.offset >= len(m.parent.buf) {
+		m.parent.cond.Wait()
+	}
+
+	n = copy(p, m.parent.buf[m.offset:])
+	m.offset += n
+
+	if m.offset >= len(m.parent.buf) {
+		if m.parent.rdErr != nil {
+			err = m.parent.rdErr
+		} else if m.parent.closed {
+			err = io.EOF
```

Comment:
returning `n > 0 && err == io.EOF` is allowed but many implementations opt not too, make sure to remember this when implementing the reader

Reply:
Yes, I am always careful to do that, thanks.